### ISSUE TITLE
proxy-injector fails with a restricted psp

### ISSUE THREAD

**Issue by codeman9** [linkerd/linkerd2#2819]
### What is the issue?
We are noticing that the proxy-injector in our cluster is unable to successfully complete injection depending on the PSP that is in place. We are using the linkerd-cni plugin and the linkerd install command is as follows:
`linkerd install-cni --proxy-uid 12102 -l tram-linkerd` (our dest-cni-bin-dir and dest-cni-net-dir might be different than the default)
`linkerd install --ha --linkerd-cni-enabled --proxy-uid 12102 -l tram-linkerd`

### How can it be reproduced?
Set up PSP as follows:
restricted-psp
```yaml
apiVersion: extensions/v1beta1
kind: PodSecurityPolicy
metadata:
  name: restricted
spec:
  allowPrivilegeEscalation: false
  fsGroup:
    ranges:
    - max: 65535
      min: 10001
    rule: MustRunAs
  readOnlyRootFilesystem: true
  requiredDropCapabilities:
  - ALL
  runAsUser:
    rule: MustRunAsNonRoot
  seLinux:
    rule: RunAsAny
  supplementalGroups:
    ranges:
    - max: 65535
      min: 10001
    rule: MustRunAs
  volumes:
  - configMap
  - emptyDir
  - projected
  - secret
  - downwardAPI
  - persistentVolumeClaim
```

permissive-psp
```yaml
apiVersion: extensions/v1beta1
kind: PodSecurityPolicy
metadata:
  name: permissive
spec:
  allowPrivilegeEscalation: true
  allowedCapabilities:
  - '*'
  fsGroup:
    rule: RunAsAny
  hostIPC: true
  hostNetwork: true
  hostPID: true
  hostPorts:
  - max: 65535
    min: 0
  privileged: true
  runAsUser:
    rule: RunAsAny
  seLinux:
    rule: RunAsAny
  supplementalGroups:
    rule: RunAsAny
  volumes:
  - '*'
```
When you have those in place, create roles and rolebindings that use the PSPs:
```yaml
apiVersion: v1
items:
- apiVersion: rbac.authorization.k8s.io/v1
  kind: Role
  metadata:
    name: hello-psp-permissive
    namespace: hello
  rules:
  - apiGroups:
    - policy
    resourceNames:
    - permissive
    resources:
    - podsecuritypolicies
    verbs:
    - use
- apiVersion: rbac.authorization.k8s.io/v1
  kind: Role
  metadata:
    name: hello-psp-restricted
    namespace: hello
  rules:
  - apiGroups:
    - policy
    resourceNames:
    - restricted
    resources:
    - podsecuritypolicies
    verbs:
    - use
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

```yaml
apiVersion: v1
items:
- apiVersion: rbac.authorization.k8s.io/v1
  kind: RoleBinding
  metadata:
    name: hello-psp
    namespace: hello
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: Role
    name: hello-psp-restricted
  subjects:
  - kind: ServiceAccount
    name: hello
    namespace: hello
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

you can delete can recreate the above rolebinding and change the `hello-psp-restricted` to `hello-psp-permissive`

Then create a standard deployment and service:

```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: helloworld
  namespace: hello
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: helloworld
        version: v1
    spec:
      containers:
      - image: rivukis/helloworld:v6
        imagePullPolicy: IfNotPresent
        name: helloworld
        ports:
        - containerPort: 8081
        securityContext:
          runAsUser: 10002
      serviceAccountName: hello
---
apiVersion: v1
kind: Service
metadata:
  name: helloworld
  namespace: hello
  labels:
    app: helloworld
spec:
  ports:
  - name: http
    port: 8081
  selector:
    app: helloworld
```

```yaml
kind: Namespace
apiVersion: v1
metadata:
  name: hello
  annotations:
    linkerd.io/inject: enabled
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: hello
  namespace: hello
```

and observe that when `restricted` is in use, the pods don't start and the replicaset has an error: `replicaset-controller  Error creating: pods "helloworld-6d8dd666c-flv47" is forbidden: unable to validate against any pod security policy: []` and when `permissive` is in use, all is well.

### Logs, error output, etc

(If the output is long, please create a [gist](https://gist.github.com/) and
paste the link here.)

#### `linkerd check` output

```text
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist

linkerd-existence
-----------------
√ control plane components ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ no invalid service profiles

linkerd-version
---------------
√ can determine the latest version
‼ cli is up-to-date
    is running version 19.5.1 but the latest edge version is 19.5.2
    see https://linkerd.io/checks/#l5d-version-cli for hints

control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 19.5.1 but the latest edge version is 19.5.2
    see https://linkerd.io/checks/#l5d-version-control for hints
√ control plane and cli versions match

Status check results are √
```

### Environment

- Kubernetes Version: v1.11.9
- Cluster Environment: self-hosted ec2 cluster
- Host OS: CoreOS
- Linkerd version: edge-19.5.1

**Comment by codeman9** [comment]
Here is a slightly more restricted psp than permissive that works:

```yaml
apiVersion: extensions/v1beta1
kind: PodSecurityPolicy
metadata:
  name: restricted-play
spec:
  allowPrivilegeEscalation: true # <------ This must be true
  fsGroup:
    ranges:
    - max: 65535
      min: 10001
    rule: MustRunAs
  readOnlyRootFilesystem: false # <------- This must be false
#  requiredDropCapabilities: # <------ This cannot be here
#  - ALL # <------ this cannot be here
  runAsUser:
    ranges:
    - max: 65535
      min: 10001
    rule: MustRunAs
  seLinux:
    rule: RunAsAny
  supplementalGroups:
    rule: RunAsAny
  volumes:
  - configMap
  - emptyDir
  - projected
  - secret
  - downwardAPI
  - persistentVolumeClaim
```
It seems like the commented lines above have to be as you see them here for the injection to work. I've tried all combinations of those 3 properties and only when they are in this state does it work.

**Comment by ihcsim** [comment]
@codeman9 I suspect this is the same issue as https://github.com/kubernetes/kubernetes/issues/65716. AIUI, the solution is to update the proxy injection code to add `spec.containers.securityContext.allowPrivilegeEscalation: false` to the proxy sidecar and the init container, **when using the CNI setup**. 

TL;DR The psp plugin mutates the pod spec to conform to the security requirements. Since this plugin is run before the mutating webhook, the injected sidecar and init container missed out on the default security settings of the psp plugin.

I'm away this week. If you are interested in submitting a PR, here's the relevant code:

* [proxy sidecar injection](https://github.com/linkerd/linkerd2/blob/b965d0d30e31bcf85852b26a5a1533afb7b704db/pkg/inject/inject.go#L438)
* [proxy-init container injection](https://github.com/linkerd/linkerd2/blob/b965d0d30e31bcf85852b26a5a1533afb7b704db/pkg/inject/inject.go#L613-L622)

Otherwise, either I can pick it up when I'm back, or others can. Thanks!

**Comment by ihcsim** [comment]
Relevant https://github.com/linkerd/linkerd2/issues/2767

**Comment by codeman9** [comment]
@ihcsim We can take this one. I don't think we need to modify the `spec.containers.securityContext.allowPrivilegeEscalation` on the init container when using CNI though, since the init container isn't used in that scenario.

**Comment by codeman9** [comment]
Ok, I found out what needs to be added and it is exactly as I show in my comment above regarding the 3 necessary components of the psp. If I add:

```go
func (conf *ResourceConfig) injectPodSpec(patch *Patch) {
	saVolumeMount := conf.serviceAccountVolumeMount()
	cniEnabled := conf.configs.GetGlobal().GetCniEnabled()
	allowPrivilegeEscalation := !cniEnabled
	if !cniEnabled {
		conf.injectProxyInit(patch, saVolumeMount)
	}

	if conf.debugSidecar != nil {
		patch.addContainer(conf.debugSidecar)
	}

	readOnlyRootFilesystem := cniEnabled
	proxyUID := conf.proxyUID()
	sidecar := corev1.Container{
		Name:                     k8s.ProxyContainerName,
		Image:                    conf.taggedProxyImage(),
		ImagePullPolicy:          conf.proxyImagePullPolicy(),
		TerminationMessagePolicy: corev1.TerminationMessageFallbackToLogsOnError,
		SecurityContext: &corev1.SecurityContext{
			RunAsUser:                &proxyUID,
			AllowPrivilegeEscalation: &allowPrivilegeEscalation,
			ReadOnlyRootFilesystem:   &readOnlyRootFilesystem,
			Capabilities: &corev1.Capabilities{
				Drop: []corev1.Capability{corev1.Capability("ALL")},
			},
		},

...

```

then my pods start up correctly. The readOnlyRootFilesystem and the Capabilities.Drop are suspect as a general solution to me. Advice?

**Comment by alpeb** [comment]
@codeman9 thanks for looking into this. From what I've been able to gather, this is independent of whether one's using CNI or not.

To complement the explanation @ihcsim wrote above, the PSP runs in two phases:

1) Right after a pod is created, the PSP controller updates it, adding the security context directives that the pod's containers are missing, like `allowPrivilegeEscalation`.
2) After all the other mutating admission controllers are run on the pod, the PSP runs again, validating that all its constraints are satisfied for this pod.

Given we're injecting the sidecar container through a mutating webhook after the PSP controller has run, step 1 isn't run on the injected sidecar container. But step 2 is, so any default directive that the PSP controller should have added is missing and the validation will fail.

Your patch would work but only for the particular case of your PSP resource. If someone decided to have `allowPrivilegeEscalation: false` in their PSP resource, the validation would fail against the `true` value you have hard-coded.

If I'm not mistaking (happy to be proven wrong), the only way for now to use PSP with the injected sidecars is to do the injection manually in your CD system, through `linkerd inject --manual` instead of relying on the proxy auto injector. That command will output your resources with the injected sidecar. And when you send that to Kubernetes, both steps 1 and 2 will run, so the injected container will have the PSP directives added to it and the validation will pass.

**Comment by codeman9** [comment]
@alpeb The reason CNI matters is because without it, you have to have already enabled NET_ADMIN privileges in your PSP. The CNI makes it so that you do not need NET_ADMIN privileges so PSPs can be much more restrictive. #2821 has some code that works for the CNI case and leaves the non-CNI case the way it was as much as possible (it does make defaults more explicit in some cases).

**Comment by ihcsim** [comment]
@codeman9 @alpeb Thanks for the inputs. Yeah, my initial thought was to specify some sane defaults for the proxy's container security context, with `allowPrivilegeEscalation` being one of them. Without CNI, if there is a PSP that sets `allowPrivilegeEscalation: false`, the proxy-init container will fail with:

```
Warning  FailedCreate  7s (x12 over 18s)  replicaset-controller  Error creating: pods "linkerd-web-7fb668dcbf-" is forbidden: unable to validate against any pod security policy: [spec.initContainers[0].securityContext.capabilities.add: Invalid value: "NET_ADMIN": capability may not be added]
```

@codeman9 Are you currently blocked by this? Will modifying the YAML generated by `linkerd inject --manual` be an acceptable workaround for now? In light of the ongoing discussion in https://github.com/linkerd/linkerd2/issues/2767, I am hoping to chat with the team (post KubeCon) to see if we can come up with a more general solution to address all the relevant admission controllers. Thanks!

**Comment by codeman9** [comment]
@ihcsim So you are saying that without the CNI, `allowPrivilegeEscalation` needs to be `true` for the init container to function correctly because it also needs `NET_ADMIN` (that is to say that `NET_ADMIN` also requires `allowPrivilegeEscalation`).

That is exactly what my PR #2821 says/does -- it sets `allowPrivilegeEscalation = true` when CNI is not enabled. There is also something about `readOnlyRootFilesystem` needing to be `false` when CNI is not enabled that seems to matter (which you can see in the PR).

In any case, this isn't blocking me really. I can use the fork on our cluster.

**Comment by ihcsim** [comment]
This is fixed under #2911.
